File: lua/class/inherit.lua
parent = {}

function parent:new(o, value)
    o = o or {}
    setmetatable(o, self)
    self.__index = self

    o.value = value or 123
    return o
end

function parent:show()
    print(self.value)
end

new = parent:new()
new:show()


child = parent:new()

function child:new(o, value, name)
    o = o or parent:new(value)
    setmetatable(o, self)
    self.__index = self
    
    o.name = name or "shit"
    return o
end

function child:show()
    print(self.name .. " : " .. self.value)
end

new2 = child:new()
new2:show()

File: cpp/class/inherit.cpp
#include <iostream>

using namespace std;

class parent
{
public:
    void print() { cout << "shit" << endl; }
    void print(int a) { cout << a << endl; }

    virtual void show() { cout << "shit" << endl; }
    virtual void show(int a) { cout << a << endl; }
};

class child : public parent
{
public:
    using parent::print; // without this, child().print(123) is illegal

    void print() { cout << "fuck" << endl; } // this will 

    void show() override { cout << "fuck" << endl; }
};

int main()
{
    child().print(123);

    child().parent::print(123);

    return 0;
};
